quinn-proto/src/connection/mod.rs

```diff
@@ -2952,6 +2993,11 @@ impl Connection {
                 &self.config,
             )
         };
+        if let Some(report) = observed_addr {
+            if let Some(updated) = new_path.update_observed_addr_report(report) {
+                self.events.push_back(Event::ObservedAddr(updated));
+            }
+        }
```

Comment:
The problem here is that when the migration code is triggered the `last_observed_addr_report` from the previous `PathData` is lost, resulting in spurious events.  I did two things to fix that:

- `PathData::from_previous` needs to copy the `last_observed_addr_report` from the previous `PathData`.  That helps when the NAT migration detection happens.  But it is not enough since that code path is only used for IPv4 port changes.  While keeping the previous observed address is always useful since you compare port and IP before emitting it again.  (side-note: maybe the NAT detection could be improved if address discovery is enabled.  But Quinn can't solve the trust issue here so that's probably not realistic)

- Because the non-NAT-migration code path also needs to work I added an extra parameter to `PathData::new`: `last_observed_addr_report: Option<ObservedAddr>`.  That way when creating a new connection you set it to `None`, and when migrating here you clone the last `ObservedAddr` into the new path.

Reply:
repeated events at the state-machine level have been fixed in 7018c84b041052eee7ea014fa70854d9471dff78, I tested this on the example works fine. As an extra measure, there is 9009d01e9f92292cda5eb108454597d56228bc37 as well

Reply:
oh yeah, so much a better fix than what I did!